Proposal for package grouping to enable release staging and release deprecation (#12874)
NuGet/Home: proposed/2023/release-staging-and-deprecation.md

   At this point, the stage for the `net8.0` group will be empty.
3. Later, as new builds are being prepared for a patch release, a new set of packages will be staged, to be later published to the same group.
As there can be multiple candidate builds for a release, there is a need for functionality to discard a stage for a given group `nuget stage --discard --group-id "net8.0"`.
This will remove all the staged but not published packages from the net8.0 group.
4. A new set of packages will be staged for the 8.0 group, untile the final build is known. At this point, there is only one unified set of related packages is staged.
5. On the release date, the release team will call `nuget publish --group-id "net8.0"` again, and only the staged packages will be published, resulting in an expanded set of published packages in the net8.0 group.
6. Some time in the future, when the release is already out of support, somebody from the .NET team will call
 `nuget deprecate --group-id "net8.0"` and all the packages which have ever been published to that group will be deprecated.
 Those packages, which has already been deprecated because of whatever reason, will not be altered.

> #### Asynchronous behavior

Author: I know I haven't called out how these async operations will be tracked. I'll work on this and include details about this too.

### .NET Release publishing

The .NET team has a monthly release cadance for their products. As part of each release we publish hundreds of packages to NuGet.
This is generally a very time-sensitive effort, especially when it comes to security releases. Many teams within Microsoft
 wait for our releases before they can move forward with publishing their patches immediately after, because with security releases
 we will also document the security fixes the release contains. Unfortunately, what happens sometimes, is that
 during publishing NuGet sometimes blocks certain packages due to some new validation rules or some other reason, which
 wasn't something the team could have foreseen earlier. As a result, the the releases get delayed and partner teams delay
 their releases awaiting a green light from the .NET team, which results in a tense / stressfull effort of patching the impacting
 packages on the last minute to address the faced issues.
Besides the security releases, there are also GA releases, which have a similar risk characteristics.

>  Many times there are confrerences schedueld the same day when a release is planned. And during such a conference the presenter

Reviewer: ```suggestion
Many times there are conferences scheduled the same day when a release is planned. And during such a conference the presenter
 expectes the release to be published (a few hours earlier) so that they can demo some new feature(s).
 And it happened in the past that the releases didn't succeed in time and impacted the presentations in a negative way.
```

# ***Release staging and deprecation***
<!-- Replace `Title` with an appropriate title for your design -->

- Author Name Artak Mkrtchyan ([mkArtakMSFT](https://github.com/mkArtakMSFT))
- GitHub Issue <https://github.com/NuGet/NuGetGallery/issues/3931>

## Summary

> This proposal introduces a concept for grouping packages that are published to NuGet.

Reviewer 2: First of all thanks for bringing up this topic and working on the spec. This looks pretty thought through including multiple scenarios. And big +1 for making this out of the package metadata, which looked like a misfit for this usecase. Most of my subsequent comments might look towards simplifying the experience for MVP. So do bear with me 😄

This will allow one to easily identify all the packages released as part of a single group as well as request bulk-processing
 for all packages which belong to the same group. For example, all packages published as part of .NET 8 release will be staged
 together and then also can be deprecated together by a single call to NuGet APIs.

## Motivation

<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
Many products (Apps / Tools / Frameworks / Utilities / etc. ...) consist of multiple interdependent packages which should be published together as a single set.

>  Currently, NuGet provides no functionality for definding / grouping packages into such sets.

Reviewer 3: ```suggestion
 Currently, NuGet provides no functionality for defining / grouping packages into such sets.
```

 Without having a concept of the `release` developers have to guess which versions of those packages are compatible.
 As a result, most of the time developers would simply pick the latest version of all these packages, which sometimes can lead to incompatibility issues down the road.
3. Keeping a package healthy on NuGet also means keeping the list of packages aligned with it support policy.
Hence, package owners try to deprecate packages which are out of support. Without ability to group packages,
 package owners end up keeping track of all the packages which belong to the same release and run the deprecation over a discrete set of packages when necessary. Being able to group packages together would enable deprecating a whole release at once, knowing that all the packages which belong to that release will get deprecated.

## Explanation

### Functional explanation

<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->

> A new NuGet CLI command `stage` will be added, which will allow a user to stage a package to nuget, specifying a `group` parameter.

Reviewer 2: Could `stage` be just an option for existing `push` command? like `dotnet nuget push  my.nupkg --stage --group mygroup`

Author: I'm afraid of the risk here. Imagine a person tries to stage a package, but accidentally mistypes the command and ends up actually making it available to the public.
This risk is not small because of two reasons:
1. `staging` should be treated as a disposable concept. So most of the packages that will be staged, will eventually be removed - freeing up space from NuGet servers, leaving no trace of these packages on the server. Because staging for a release will happen in groups, imagine accidentally not staging, but publishing a whole set of packages to NuGet. Removing packages, as I understand, is not possible today (unlisting is, but it's not the same).
2. What if a package which was meant to be staged only, contains a security patch?

Reviewer 4: Those are very valid concerns! In the end, I still feel like staging is part of the push process, rather than something separate.

Reviewer 3: Currently symbol push and nupkg push are both lumped under `dotnet nuget push` so it seems like that's the place to expand the "push" concept rather than adding a new verb (i.e. we didn't add a `dotnet nuget symbol-push` at the time, we overloaded the existing concept. 

That being said, it's under the `nuget` so having `dotnet nuget stage` doesn't seem so bad either. 

I think having it as part of `push` is a bit better since all of the options on `push` right now probably make sense for `stage` (see `dotnet nuget push --help`). So if we had a separate verb we'd be copying all of those options over or having to make the decision of whether or not to copy it over. Easier to overload IMO.

Reviewer 3: Another point is that the `dotnet nuget delete` gesture (the buddy of `dotnet nuget push`) would need another new verb like `unstage` to pair with `stage`. At that point it feels like too many new verbs.

Reviewer 3: It might be worth alluding to the fact that not all sources will implement staging immediately or every. The `--stage` option (or `stage` verb) should fail gracefully when stage isn't supported. At the implementation level this means adding a way for sources to advertise staging support, which is a solve problem with our V3 service index model -- servers can advertise capabilities and client can enrich the error/fallback experience in that case.

Author: >Another point is that the dotnet nuget delete gesture (the buddy of dotnet nuget push) would need another new verb like unstage to pair with stage. At that point it feels like too many new verbs.

Alternatively, the --stage option can be applicable to the `dotnet nuget delete` option, so there is no extra word involved.

 Staged packages will be uploaded to NuGet and will be validated during the staging process, so that the staging fails

>  if validation fails for a package in a group. After uploading multiple packages, the user user can then call `nuget publish [group]`

Reviewer 2: How would publisher know if some packages failed tests and hence need to be re-staged? Is this command going to be blocking? Do we need a command option to get status? Or will there be a notification (email or otherwise)?

Author: The validation is going to happen during the staging phase. And I have a whole section (Asynchronous behavior) about the need for status reporting and the need for a operation status command (line 126)

Reviewer 4: I'm wondering if adding a publish command becomes confusing, given that push has always been the equivalent to publish in nuget.

If I have both, how do I decide which one to use as a new user?

Author: Sorry, I didn't mean to introduce a new way for pushing. I meant to still use the `nuget push` command here, just with an additional `group` parameter.

>  if validation fails for a package in a group. After uploading multiple packages, the user user can then call `nuget publish [group]`

Author: ```suggestion
 if validation fails for a package in a group. After uploading multiple packages, the user user can then call `nuget push --group [group]`
```

>  if validation fails for a package in a group. After uploading multiple packages, the user user can then call `nuget push --group [group]`

Reviewer 3: Below this command is `nuget publish` not `nuget push`. I think overloading `nuget push` for a broad group publishing gesture is more questionable that the conversation here: https://github.com/NuGet/Home/pull/12874/files#r1379170155.

there's two gesture types: operation on individual nupkgs (pushing to a source directly or to a stage) and group operations (operating not on an individual nupkg but on the whole group). I think `push` can do it all or perhaps just the nupkg level operations. The group level operations could potentially have their own verb, but this is up to the client team to decide.

Author: Updated to using nuget push and nuget delete commands as the main ones used for managing stages / staged packages.

>  command to request all the packages from the earlier group to become availble on the server as `published`.

Reviewer 3: ```suggestion
 command to request all the packages from the earlier group to become available on the server as `published`.
```

Note, that the `group` information will only be stored on the server and has nothing to do with how a package was built.

> Groups are scoped to an owner. So a package published by `Owner 1` into `Group 1` has nothing to do with another package published

Reviewer 2: What happens when both owners have permissions to push the package? _(Thinking of multiple owners being part of one or many org accounts on nuget.org.)_

Author: It doesn't really matter how many. If a person has access to push a package, they can perform that operation.

 to `Group 1` by `Owner 2`, as these should be treated as two separate groups.

Groups can expand over time, by publishing new packages to the same group in the future. There is no concept of a "deprecated group"
 as group by itself is stateless. It's just a collection of packages. A package can belong only to a single group.
 This group expansion functionality will be handy for the error situation described above, where during staging a package staging fails.
 This will result in only some packages being staged. The user will need to address the issues related to the failed package

>  and stage the remaining packages using the same command. Note, that if a package is stages, the operation should be idempotent,

Reviewer 3: ```suggestion
 and stage the remaining packages using the same command. Note, that if a package is staged, the operation should be idempotent,
```

 and a request to stage the same package again should only result in a warning message (package is already staged) and the package shouldn't be uploaded again.

Groups can later be used in the future when it's time to deprecate a set of packages. The package author can request to
 deprecate a `group` by calling a single command on the nuget CLI and all the packages that belong to that group will be marked as `deprecated`.

With this information in place, new set of NuGet APIs / extensions will be needed to allow a user to run the following actions:
- **Stage** a package for later publishing
- **Publish an earlier staged release**
- **Deprecate a group** using the group id 
- **Discard a staged release** using the group id

Note, that the user is required to be the owner / co-owner of the packages which are in the group,
otherwise the request to make any modification to a non-owned package will fail.

#### Package staging
When a package is staged, the id of the package is reserved at that point, and no other user can introduce a package with the same name.

> Also, to avoid abuse, only a limited number of packages can be staged within a single group. While this can potentially be defined as a configurable value per organization / owner, there is no such requirement at the moment. Hence, this can be set to 1000 packages for now or not defined at all. Will leave this for potentially future consideration.

Reviewer 3: Does 1000 work for .NET releases? I think this is an important bound to understand sooner rather than later since impacts the viability of a "stage publish SLO" (how long it takes to publish up to X packages quickly)

Author: I checked and for .NET 8 release we had a bit less than 700 packages. So 1000 is a reasonable default cap, I think.

> When staged, packages will not be visible / accessible to anyone but the owner (including coowners). They will be able to see the current staged list for a given group using `nuget stage --list --group-id [group-id]` command. 

Reviewer 2: Okay, you answered one of the questions I had in my previous comment 😄. Is nuget.org UX planned for this? Ideally both need to be in sync.

Author: No ask for a NuGet UI experience at the moment. It would be great, but I don't think it's important for the MVP.

Reviewer 3: UI may be easier for us to ship than a CLI experience on a case by case basis. For each new CLI gesture, consider the alternative of a UI and propose what you believe is the minimal for the experience to work (or say either UI or CLI works). For example this list experience may be fine just as UI saving us one more place where there needs to be a new protocol between client and server.

> When staged, packages will not be visible / accessible to anyone but the owner (including coowners). They will be able to see the current staged list for a given group using `nuget stage --list --group-id [group-id]` command. 

Reviewer 3: ```suggestion
When staged, packages will not be visible / accessible to anyone but the co-owners. They will be able to see the current staged list for a given group using `nuget stage --list --group-id [group-id]` command. 
```

The user that pushed the package should never have different powers than a co-owner. Our design of co-ownership is complete trust and complete, mutual ownership.

#### Group Ownership
There are few interesting scenarios related to package ownership:
1. The user has staged a package to a group for later publishing, but the ownership has changed, and the user is no longer
 an owner for a package in a group. At this point, publishing request should fail with a message that the group has been modified
 from its original set, and the outcome of the request may be different from what the user was expecting, indicating the package
 that was moved out of the group. If the user is aware of the problem and still wants to move forward with the change,
 they will have to pass an additional `--force` parameter to the publish command: `nuget publish <group-id> --force`.
2. The group to be published no longer has any packages in it. In this case the command should fail detailing the reason for the failure.
3. This will be the same but for deprecation. Essentially, any bulk operation using groups, where the set of the packages in the group
 have changed after staging, should result in an error with details about the change. The `--force` flag should again be used

>  to ignore the change and move forward with execution of the requst (either publishing or deprecation). To summarize,

Reviewer 3: ```suggestion
 to ignore the change and move forward with execution of the request (either publishing or deprecation). To summarize,
```

 the fundamental principle is that a user cannot deprecate a package they don't own.

> #### Lifespan of a staged group

Reviewer 2: To simplify this, could there be a set timeline for this feature. Say 1 month? After which these packages just get deleted if in staged mode?

Author: I think that's what I have specificed already, haven't I:

> there should be a maximum lifespan for a staged package, that is not published. NuGet should then unstage / remove any package that was staged but never published within the specified / pre-configured time period. As a maximum timeslot a default one month period can be considered.

Maybe I'm misunderstanding you. Or do you mean to not even give the users ability to specify any expiration period, and only stick with the "maximum" one from the server?

Reviewer 2: yes. For MVP, a default expiration should work for all.

Author: Got it. I think it will make sense to explicitly call out that the "customized expiration" should be treated as a P2 for the eng. team to consider when they can. I'll make a note of that.

If users keep staging packages but do nothing with them, NuGet will become an ever-growing package graveyard,
 incurring higher and higher costs over time simply for storing these packages. To avoid this problem, staging should be time-limited.
 That is, there should be a maximum lifespan for a staged package, that is not published. NuGet should then unstage / remove
 any package that was staged but never published within the specified / pre-configured time period. As a maximum timeslot
 a default one month period can be considered. However, to reduce the cost of maintaining these packages,
 the `nuget stage` command can also accept an optional `--lifespan [number of days]` parameter,
 which can indicate a shorter time period, thus giving NuGet ability to remove those packages earlier than the maximum period.

> Note, that this opens up a new scenario, where a group has been created with a package 29 days ago, and then a new package is staged to the same group.

Reviewer 3: This essentially mean the TTL lives on the stage package not the group, if I understand correctly. 

I think we can make this simpler and just keep the TTL on the stage itself. When any package gets pushed, the TTL gets reset. An owner could keep the package alive forever by removing then adding the same package every day. That's fine. They can come to the same effect with by deleting it entirely and creating it every day with CLI commands. It's all scriptable since we will be having CLI not UI gestures.

In light of that, I think a simpler implementation is fine.

Author: Agreed. I've simplified this section based on your suggestion. Thanks!

In two days from that time, the package author tries to publish the staged group. At this point, the publish command will fail,
indicating that the group has been modified, and showing the package that has been deleted by NuGet out of the staging due to the expiration.
The user can still continue publishing using the `--force` option, which will result in the non-expired/ removed staged packages to be published.
Alternatively, the user can restage the expired package and call the publish command again.

> #### Expanding groups

Reviewer 2: Not sure if you answered these in the spec already, but some questions:
1. Can one put multiple versions of the same package ID into the group?
2. Will normal push (publishing as it exists today without staging) of same version allowed, if that version exists in the staged group?
3. What happens if a later version is published directly than what a staged group has? Will publishing of a newer version impact the staged group in anyway?
4. Can one member of org stage the group of packages and another member of the org publish? Will this require `--force` option?
5. What happens to the package ID namespace if a staged group is discarded? Same question for package version? Are they released for all?

Author: > 1. Can one put multiple versions of the same package ID into the group?

I was thinking that the `Package ID` includes the version too. This will mean that the answer is "yes, multiple versions of the same package can be staged at the same time"

> 2. Will normal push (publishing as it exists today without staging) of same version allowed, if that version exists in the staged group?

Because server validation will now happen during staging and if a package was staged successfully, NuGet should guarantee that it can be published, this means that staging will "reserve" package ID (including version). Thus, if a package was already staged, that version / id can't be published.

> 3. What happens if a later version is published directly than what a staged group has? Will publishing of a newer version impact the staged group in anyway?

I believe the answer above covers this too. If a package is staged, publishing it directly won't be possible. Also, if a package is already published, staging the same (id / version ) will fail.

> 4. Can one member of org stage the group of packages and another member of the org publish? Will this require --force option?

Yes, a member can publish packages even if they were staged by another user (as long as they both have the necessary permissions to publish each individual package). The --force option is irrelevant here.

> 5. What happens to the package ID namespace if a staged group is discarded? Same question for package version? Are they released for all?

Correct. If a stage is discarded the "reservation" for the package ID will no longer exist, so anybody can claim it again.

It should be possible for a group to expand over time. This is currently a real scenario, where .NET GA SDK ships a set of packages,
 and then follow-up SDK updates ship updated packages which still belong to the same .NET 8 package group.
 Here how this will play out with the tooling support described above:

> 1. During GA release, the release team will stage the set of RTM packages for the release. (nuget stage --group-id "net8.0" --package-id <package-id>)

Reviewer 3: Why would `package-id` be specified here, not a path to a .nupkg?

Author: You're right. This should be the path to the package.

Author: ```suggestion
1. During GA release, the release team will stage the set of RTM packages for the release. (nuget stage --group-id "net8.0" --package-id <package-file-path>)
```

2. Then on the day of the release, all the staged packages will be published using `nuget publish --group-id "net8.0"` command.
   At this point, the stage for the `net8.0` group will be empty.
3. Later, as new builds are being prepared for a patch release, a new set of packages will be staged, to be later published to the same group.

> As there can be multiple candidate builds for a release, there is a need for functionality to discard a stage for a given group `nuget stage --discard --group-id "net8.0"`.

Reviewer 3: I think it's a little inconsistent to have the `discard` verb hidden as an option `--discard`. It almost feels like we want `dotnet nuget stage discard` or (perhaps more aligned with .NET CLI standards) `dotnet nuget remove stage`. I'm not an expert here but it seems inconsistent with current .NET CLI or even `dotnet nuget --help` patterns.

Author: Updated to us nuget delete command option instead.

This will remove all the staged but not published packages from the net8.0 group.

> 4. A new set of packages will be staged for the 8.0 group, untile the final build is known. At this point, there is only one unified set of related packages is staged.

Reviewer 3: ```suggestion
4. A new set of packages will be staged for the 8.0 group, until the final build is known. At this point, there is only one unified set of related packages is staged.
```

5. On the release date, the release team will call `nuget publish --group-id "net8.0"` again, and only the staged packages will be published, resulting in an expanded set of published packages in the net8.0 group.
6. Some time in the future, when the release is already out of support, somebody from the .NET team will call

>  `nuget deprecate --group-id "net8.0"` and all the packages which have ever been published to that group will be deprecated.

Reviewer 3: There is absolutely no CLI support for deprecate today :(. As you know it's a private preview API with no CLI support. I think this part of the spec is blocked on the public preview preview of deprecate API and perhaps corresponding non-stage-related CLI deprecate commands.

 Those packages, which has already been deprecated because of whatever reason, will not be altered.

#### Asynchronous behavior

As both publishing and deprecation commands can take some time, these will run asynchronously.
 So the CLI command for both operations will return immediately, giving the user some token which can later be used to
 track the current state of the request.
There are no critical scenarios for deprecation to happen within specific time period, however, publishing is somewhat
 critical and is being orchestrated in many cases. Hence, NuGet should provide some guarantees for how much time the operation
 can take in the worst case. As of right now, having 10 minutes deadline for the publishing request should be reasonable.

> This requirement brings a need for a new NuGet command for querying the status of an asynchronous operation.

Author: ```suggestion
This requirement brings a need for a new NuGet command for querying the status of an asynchronous operation. This particular requirement should be treated as a stretch goal, and it can be implemented in a later release.
```

Let's assume that the group publishing and deprecation commands will return some "operation id" for later state inquriry by the client.

> `nuget status --operation-id` command can be used to track the progress of an asynchronous operation.

Reviewer 3: Makes sense, but I think this can be put as a stretch goal or secondary priority. This, like listing the contents of a stage, may be easier implemented in UI for now.

> This can return the fullowing results:

Reviewer 3: ```suggestion
This can return the following results:
```

# ***Release staging and deprecation***
<!-- Replace `Title` with an appropriate title for your design -->

- Author Name Artak Mkrtchyan ([mkArtakMSFT](https://github.com/mkArtakMSFT))

> - GitHub Issue <!-- GitHub Issue link -->

Reviewer 4: ```suggestion
- GitHub Issue <https://github.com/NuGet/NuGetGallery/issues/3931>
```

## Summary

> This proposal introduces a concept of a logical grouping for packages published to NuGet. This will allow one to easily identify all the packages released as part of a single product version release as well as request bulk-processing for a set of packages which belong to the same release. For example, all packages published as part of .NET 8 Preview 5 release will be staged together and then also can be deprecated together by a single call to NuGet APIs.

Reviewer 5: Markdown allows you to put each sentence on a new line in "the code", but when rendered it still appears in one paragraph (I never looked at the rendered HTML, I suspect it's more a feature of HTML, than of Markdown). Anyway, doing so makes it easier to respond to individual sentences, rather than a paragraph as a whole.

```suggestion
This proposal introduces a concept of a logical grouping for packages published to NuGet.
This will allow one to easily identify all the packages released as part of a single product version release as well as request bulk-processing for a set of packages which belong to the same release.
For example, all packages published as part of .NET 8 Preview 5 release will be staged together and then also can be deprecated together by a single call to NuGet APIs.
```

## Motivation 

<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
Many products (Apps / Tools / Frameworks / Utilities / etc. ...) consist of multiple interdependent packages which should be published together as a single set. Currently, NuGet provides no functionality for definding / grouping packages into such sets. This results in package owners having to keep track of these package sets on their own. There are three specific sets of scenarios which this feature will simplify dramatically:

> 1. When preparing a release, product owners would like to have a mechanism to pre-stage their product packages, so that the release will consist only of a single action of marking the set of packages as `publishged`, which would result all the packages in the set to be marked as published at the same time.

Reviewer 5: ```suggestion
1. When preparing a release, product owners would like to have a mechanism to pre-stage their product packages, so that the release will consist only of a single action of marking the set of packages as `published`, which would result all the packages in the set to be marked as published at the same time.
```

3. Keeping a package healthy on NuGet also means keeping the list of packages aligned with it support policy. Hence, package owners try to deprecate packages which are out of support. Without ability to group packages, package owners end up keeping track of all the packages which belong to the same release and run the deprecation over a discrete set of packages when necessary. Being able to group packages together would enable deprecating a whole release at once, knowing that all the packages which belong to that release will get deprecated.

## Explanation

### Functional explanation

<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->

> There is a new field in the package metadata, which indicates the unique release of the product the package belongs to. This field is called `release` and the value is scoped to the publisher of the package.

Reviewer 4: You mention a tag is unrestricted, but doesn't this new metadata have the exact same problem?

Author: Actually, I think there is a need for restricting this new metadata and validating that it's not something already used.
This will give release publishers a peace of mind that the release is unique (for the publisher/author) and that packages can't get mixed up.

Reviewer 4: It'd be great to add that somewhere in the proposal.

Author: This is now covered in the proposal.

> When a package is being built, that field in the metadata is set to the same value for all the related packages, that are built together.

Reviewer 5: Assuming that the .NET team end up using a release named ".NET 8.0.1", and later ".NET 8.0.2", what happens if some random person, not in the Microsoft or dotnet orgs, pushes a package using one of these release names?  What happens if someone not from Microsoft publishes a package using the ".NET 8.0.3" release name before Microsoft/dotnet does?

Author: That's exactly why I mentioned [here]() that the release should be scoped to the publisher. So that anyone else from another organization can have a release named exactly the same way and those will not interfere with each other by any means.

> Now, with this information in place, new set of NuGet APIs / extensions allow the user to run the following actions:

Reviewer 4: A lot of these actions are most appropriate if the published/org is the same and putting that information into the package metadata is IMO too early in the process. 

What if you try to "deprecate" packages that have the same "release", but they're not owned by the same org. 

To eliminate that confusion, how about making this a server only/publisher feature? So you'd be able to push any package as `staging`. Then the server will be able to finesse the UX around it as needed, something that can evolve over time.

Author: Well, on the other hand, this information is very similar to the version, which is something known at build time.

As for the following, I don't think this is a valid scenario. Those should be two separate deprecation sessions.
> What if you try to "deprecate" packages that have the same "release", but they're not owned by the same org.

>how about making this a server only/publisher feature? So you'd be able to push any package as staging. Then the server will be able to finesse the UX around it as needed, something that can evolve over time.

Can you please elaborate on this? Are you suggesting that the server should "figure out" the group to assign the staged package to? How would the server do that without any explicit information?

Reviewer 5: The server can generate a batch ID though some API, and nupkgs being uploaded will need to pass that batch ID as an HTTP header when pushing individual nupkgs. Whatever tool does the uploads (if all this grouped release/deprecation features are nuget.org only, I will argue we should then not build it into `dotnet nuget push`) can automate all that complexity when `push --batch *.nupkg` is run.

Your proposal to put the release name in the nuspec allows packages to be uploaded independently more easily. But even with the complexity of the .NET build, my understandings is that an Azure Artifacts feed is created with a GUID or some other generated name, and each part of the build pushes to there. Therefore, I believe it wouldn't be unreasonably difficult to download all the packages from this feed onto a CI agent, and then do the `push *.nupkg` to nuget.org, so that it is done in one transaction.

I think this idea of server generated ID should be described in the document. If it doesn't end up being the accepted design, then under the "alternate designs" section with an explanation of its drawbacks. Otherwise this "package author puts release in nuspec" should be in the alternate design feature.

Honestly, my mind races with security related questions when the package author defines the release id/name. Are different nuget.org accounts allowed to use the same name? Or does the feature enforce each account to have a different "release namespace" per account?  How would that work with something like Azure Artifacts, or [other servers](https://learn.microsoft.com/en-us/nuget/hosting-packages/overview) that implement the nuget server API?  I'm sure there are more risks if we think about it. My gut feeling is that allowing the server to generate the ID makes security easier.

Reviewer 4: The way I was thinking about it is slightly different from @zivkan's. 

My thinking is that we can just have a grouping that's configured in any way shape of work *on* the server itself. 
For example, the server could choose how they represent releases and the packages are configured there. The client *doesn't* need to know that. 

The client can imo know 2 things. 
1. Push as unlisted/staged?
1. Move from staged to "available" or whatever we describe the verb is. 

The client then wouldn't need to know of any new metadata/release id etc. 

IMO by not having this metadata *in* the package itself, we avoid chicken/egg problems of build time vs branding. We don't have any of the ownership questions, since theoretically a `release` could cover multiple owners. 

The disadvantage of this idea is that it'd be impossible to query for these packages locally.

Author: Thanks for sharing your thoughts / ideas @zivkan, @nkolev92.
I don't think letting the server decide what the "group" is correct. What if for whatever reason the first batch-upload session got interrupted and now I need to have another session to upload the remaining packages? Or even without interruptions, I just happen to upload the packages which are part of the same release in several sessions.

I just keep coming to the current experience. We do know the version / release the package belongs to when we build it.
Maybe @leecow can share more information about if we ever have a situation, where the branding will be defined after the packages are built?

Reviewer 4: I'm not necessarily referring to the server requiring things to come in one upload. 
It'd be something that can be "connected" independently, regardless of who it was upload/what metadata it has. 

It can be something similar to how API keys are done, for example when you create API keys on the server, you can create scopes for individual package ids. This could work similarly.

Reviewer 5: I think calling it not correct is a bit strong. Your proposal of having the package author name the release whatever they want, and forcing the server to partition the labels by account/whatever is imposing a specific design on server implementations. Maybe that's fine, maybe that has performance risks due to the server's existing implementation. We need input from the server team, and unfortunately we've only had comments from NuGet's client team so far.

In particular, nuget.org's organization accounts add complexity. Say I'm publishing new versions of NuGet.* packages for release 6.8.0. [These packages are owned by the Microsoft org, the NuGet org, and I'd be pushing from my own `zivkan` account](https://www.nuget.org/packages/NuGet.Protocol). Exactly what scope is this `6.8.0` release?

If there's any other package owned by the Microsoft org account with the release label `6.8.0`, do my NuGet.* packages become part of that release?  When I tell nuget.org to "make 6.8.0 public", will I affect those other packages?

Or since I pushed the package, will "release 6.8.0" be scoped to my `zivkan` account. But that means if I stage the packages before GA day, but then I'm sick or no longer an employee, how will someone else publish it as public?  Or mark the whole release as deprecated?

What would be ideal is to be able to say the release is scoped to the NuGet org account, allowing other members of the same org to interact with it, but the Microsoft org can't see this package's release scope, despite being a co-owner of the package.

These are the types of issues I hope the server team can provide good feedback on.

Reviewer 4: > Or since I pushed the package, will "release 6.8.0" be scoped to my zivkan account. But that means if I stage the packages before GA day, but then I'm sick or no longer an employee, how will someone else publish it as public? Or mark the whole release as deprecated?

I think users do things **on behalf** of orgs. So I can "push" a package, but you should be able to change the visibility. If you're not logged in, you can't really see which individual person pushed it.

Reviewer 5: How prescriptive should feature specs be? Is this something that the spec should explicitly mention/address, or leave the spec "open", so different servers can implement it however they see fit, but therefore customers pushing packages have different experiences depending on which server they push to?

Reviewer 3: I agree with much of what @zivkan has said here. There is a decision between a pack-time release label and a server-owned batch/release ID. Both should be considered in full. I think a `batch ID` GUID is one option, but another option is to allow an explicit release entity to be created as a separate gesture and then, over time, packages can be explicitly added or removed to the release entity. This could even be defaulted based on package metadata somewhat marrying the two options.

I think it should be decided now because it has big impact on what the author workflow would be like. Some options require enhancements to `pack`, others require enhancements to `push`, others could be server-only with a `add package to release` or `stage package` gesture that perhaps need not be implemented in NuGet client immediately.

> Now, with this information in place, new set of NuGet APIs / extensions allow the user to run the following actions:

Reviewer 5: To be clear, is this proposal suggesting to change the [NuGet Server API](https://learn.microsoft.com/en-us/nuget/api/overview) and therefore all NuGet feeds can (should?) implement?  Or is the proposal for nuget.org only?

How much thought should we put into how much the feature makes sense in contexts other than nuget.org, such as [the many servers listed in our docs on hosting your own feeds](https://learn.microsoft.com/en-us/nuget/hosting-packages/overview). Some of the servers seem more tailored to hosting a single feed, similar to nuget.org, whereas others, particularly the SaaS offerings, seem to be more tailored towards an integrated CI/CD or DevOps environment

Reviewer 3: Any API changes should be documented on the https://learn.microsoft.com/en-us/nuget/api/overview now, or perhaps eventually, instead of staying on a page like https://learn.microsoft.com/en-us/nuget/api/nuget-protocols forever. The NuGet team is steward to the NuGet ecosystem which intersects with concerns from other feed implementors. 

For the case of the deprecation API, NuGet.org is the first to implement it and it's in a private preview state only on NuGet.org, but when it graduates to public preview it will be documented on the NuGet Server API much like the push or unlist/delete API so other feeds can pick it up if they want.

> - **Search / Query for packages based on a release** for a given publisher

Reviewer 5: The motivation explains why being able to "batch" nupkgs for publishing, and maybe deprecating the whole lot as a group is important. However, I feel like this search request isn't justified from a customer experience point of view and may be feature creep.

Let's say nuget.org has adds "Package Batches" to https://www.nuget.org/account/Packages so you can see your own batches, is that sufficient?  Why would a 3rd party want to, or need to, see which packages are included in a "release"?

This is related to another comment that I, and Nikolche, make about not putting the release name/label in the nuspec. Especially around security issues like namespace clashes and typosquatting.

Author: I have mentioned this, but I agree that the search is not a critical / required aspect of this proposal. I'd be happy to exclude it from the list.

> - **Deprecate a relase** for current publisher

Reviewer 5: Maybe my questions are more suitable for a "technical spec", rather than a "user spec", but this is how my brain works. I can't help but think about all the little details as if I was being asked to implement it immediately.

With the current proposal, where the package author decides the release name before creating their package, what happens if a package is pushed to a release after the release has already been marked as deprecated? Should the new package be immediately marked as deprecated as well? Should a release become "locked" when marked as deprecated, returning an error is any new package is uploaded with the same release?  Should it be server implementation dependent?

Author: I agree this is too technical. From the UX standpoint that doesn't make sense. If the release is deprecated I'd expect no further modifications to it should be made. So I'd say this should result in an error.

3. Keeping a package healthy on NuGet also means keeping the list of packages aligned with it support policy. Hence, package owners try to deprecate packages which are out of support. Without ability to group packages, package owners end up keeping track of all the packages which belong to the same release and run the deprecation over a discrete set of packages when necessary. Being able to group packages together would enable deprecating a whole release at once, knowing that all the packages which belong to that release will get deprecated.

## Explanation

### Functional explanation

<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->

> There is a new field in the package metadata, which indicates the unique release of the product the package belongs to. This field is called `release` and the value is scoped to the publisher of the package.

Reviewer 5: Markdown and/or HTML allows paragraphs to be split into multiple markdown lines, which makes commenting on specific parts within a paragraph easier.  I'm also suggesting a change to the second sentence, to help clarify something I missed in my first read though.

```suggestion
There is a new field in the package metadata named `release`, which indicates the unique release of the product the package belongs to.
Server implementations are expected to treat each `release` as scoped to an account or similar concept, depending on the server's implementation details.
For example, on nuget.org, User1 and User2 can both use a `release` named `V1.0.0`, but when User1 operates on their `V1.0.0` release, it does not affect User2's release with the same name.
```

When a package is being built, that field in the metadata is set to the same value for all the related packages, that are built together.

Now, with this information in place, new set of NuGet APIs / extensions allow the user to run the following actions:
- **Search / Query for packages based on a release** for a given publisher
- **Deprecate a relase** for current publisher
- **Stage** a package for later publishing

> - **Publish an earlier staged release**

Reviewer 5: Say a package ID that has never been published before is pre-staged. At what point should the the package ID be claimed, preventing any other account from pubishing a package with the same ID?  Or should each server be free to implement whatever they feel is best?

Author: I assume what you refer here is not package ID, but the new metadata I introduced - "release".
I'm having hard time understand the concept of different `servers`. Maybe I lack some critical knowledge here. I'll schedule a quick meeting to go over that with you and maybe then I'll be able to better answer this question.

For now, I assume something like this can work:
1. Staging request on its own can be that signal. So that when the publisher is about to upload a set of packages for staging, they first call some `stage release {ReleaseID}` API, passing in the release they'd like to use.
2. They receive back some "key" in response to the call above
3. For each package upload they provide the key as a way to associate the package with the release.

This addresses the issue of somebody else using the same release to upload another package, because they don't have the unique key, which will be considered a temporary secret.

Reviewer 5: > I assume what you refer here is not package ID, but the new metadata I introduced - "release".

No I mean package ID.  Here's the scenario I'm thinking about:

User1 creates a package UsefulUtils, which we'll pretend does not yet exist on nuget.org, and stages it, waiting until their official product announcement & release date. Before User1 presses the button to make the package public, User2 pushes a package, with the same ID, UsefulUtils.  Later, User1 tries to publish their staged release.  Only one of these two can succeed. On servers like Azure Artifacts where anyone who can push to a feed has access to all package IDs, if User1 and User2 push different versions, then it will work. On servers like nuget.org, where package ID have their own ACL, one of the two users will not be able to successfully publish their package, unless they're both part of the same org account, and the package is published under that org account.

Should this spec be prescriptive about the expected behaviour? (we have [docs on the NuGet Server API](https://learn.microsoft.com/en-us/nuget/api/overview) that tell anyone creating their own NuGet server how it's expected to behave) Or should this be left up to each server to make their own decisions, leading to potential inconsistent behaviours from the point of view of package authors?

> I'm having hard time understand the concept of different servers.

We have [18 different NuGet server implementations listed in our docs](https://learn.microsoft.com/en-us/nuget/hosting-packages/overview).

While nuget.org is of course a very, very important server implementation to the NuGet and .NET ecosystem, I personally hold a strong view that we should not design or implement features that are hyper-specific to one server, even nuget.org. It's fine with me if nuget.org implements features that we don't intend on other servers to implement, but then I don't believe that `dotnet nuget` commands should have commands that only work with that feature, even nuget.org. If we want to add commands into the dotnet CLI that interact with nuget.org, then the feature needs to be designed to allow any other server to reasonably implement it.

Reviewer 3: At what point the package ID is an important question, but I think it's easy. It should be reserved when the package enters the staging area, not later when the staging area is published as a batch. Any concerns of locking up too many package IDs in the stage should be mitigated with throttling or rate-limiting activities done automatically today or handled on a case-by-case basis (many abuse/fraud cases are special and need human eyes to mitigate). 

Having the package ID reserved only later would reduce the usefulness of the staging area. There should be as few reasons as possible for the staging area to fail publishing. Package metadata validation, signing validation, ownership checks, etc etc all should happen as early as possible so the publish of a staging area (i.e. release) is a smooth operation.

Author: Agreed with @joelverhagen here, and making a note in the proposal about this.

> - **Publish an earlier staged release**

Reviewer 5: I'm wondering if there's potential for issues around two different people who are co-owners of a package ID from staging the same version of a package. What should the semantics be?  Or leave it as undefined in the feature spec and different servers can handle it in different ways?

Author: Shared an idea in the comment above.

Author: As per this conversation (see https://github.com/NuGet/Home/pull/12874/files#r1333197815), package IDs will be reserved as soon as staged, so nobody else should be able to stage / publish the same package.

For example, all packages published as part of .NET 8 Preview 5 release will be staged together and then also can be deprecated together by a single call to NuGet APIs.

## Motivation 

<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
Many products (Apps / Tools / Frameworks / Utilities / etc. ...) consist of multiple interdependent packages which should be published together as a single set. Currently, NuGet provides no functionality for definding / grouping packages into such sets. This results in package owners having to keep track of these package sets on their own. There are three specific sets of scenarios which this feature will simplify dramatically:
1. When preparing a release, product owners would like to have a mechanism to pre-stage their product packages, so that the release will consist only of a single action of marking the set of packages as `published`, which would result all the packages in the set to be marked as published at the same time.

> 2. Sometimes developers have to take a dependency on several related packages in their projects, which are part of the same "release". Without having a concept of the `release` developers have to guess which versions of those packages are compatible. As a result, most of hte time developers would simply pick the latest version of all these packages, which somtimes can lead to incompatibility issues down the road.

Reviewer 3: ```suggestion
2. Sometimes developers have to take a dependency on several related packages in their projects, which are part of the same "release". Without having a concept of the `release` developers have to guess which versions of those packages are compatible. As a result, most of the time developers would simply pick the latest version of all these packages, which sometimes can lead to incompatibility issues down the road.
```

3. Keeping a package healthy on NuGet also means keeping the list of packages aligned with it support policy. Hence, package owners try to deprecate packages which are out of support. Without ability to group packages, package owners end up keeping track of all the packages which belong to the same release and run the deprecation over a discrete set of packages when necessary. Being able to group packages together would enable deprecating a whole release at once, knowing that all the packages which belong to that release will get deprecated.

## Explanation

> ### Functional explanation

Reviewer 3: Should symbol packages (.snupkgs) be stage-able also? If so, and we opt to use a release label that's in the .nuspec, then .snupkg pack would need to be enhanced also. I believe the .nuspec in the .snupkg is some subset of the related .nupkg .nuspec.

Author: Given that the "group" information is no longer in the package metadata, do we have to treat symbols packages differently?

Reviewer 3: Sorry, missed this response. No, they would not be different then. No metadata change but `nuget push` probably needs a gesture to "push to stage" and this would work on .nupkg or .snupkg.

<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
There is a new field in the package metadata named `release`, which indicates the unique release of the product the package belongs to.
Server implementations are expected to treat each `release` as scoped to an account or similar concept, depending on the server's implementation details.

> For example, on nuget.org, User1 and User2 can both use a `release` named `V1.0.0`, but when User1 operates on their `V1.0.0` release, it does not affect User2's release with the same name.

Reviewer 3: This concept needs to be reconciled with multiple owners on a package. I take it that the set of owners on a package containing this `release` all share the same `release` entity? Theoretically only the entity that pushed the specific package version (i.e. the organization of the pusher or the specific co-owner that pushed the package) could relate to the `release` entity. This is a big ambiguous of a point I'm making here, so let's take an example.

If user `joelverhagen` pushes release `v1.0.0` to his `Knapcode.TorSharp` package but `joelverhagen` co-owns a package `Contoso.MyLib` with `artak`, does Artak share the `v1.0.0` release entity? Probably not. But what if `v1.0.0` was used on `Contoso.MyLib`, the shared package. Then `joelverhagen` now has two potentially unrelated `v1.0.0` release entities. 

Mutli-owner packages introduce these complexities.

Author: This is indeed the most challenging part of the proposal. Including in the updated one that I'm working on now. Still trying to come up with a good plan here.

When a package is being built, that field in the metadata is set to the same value for all the related packages, that are built together.

Now, with this information in place, new set of NuGet APIs / extensions allow the user to run the following actions:

> - **Stage** a package for later publishing

Reviewer 3: What is an acceptable level of visibility for staged packages? I am exploring this because it has real impact on implementation cost.

Suppose a staged package is simply unlisted. It's findable by an unauthenticated user who knows the version number but does not appear in package search or version lists in VS or NuGet.org UI. This is a cheaper thing for NuGet.org to implement. Publishing the release/staged packages would simply be marking N packages as listed.

If we need to do a mega, big bang release of hundreds of totally unpublished, staged packages, this can impact the availability of our publishing pipeline. Today, there is a relationship between package push rate limits and throughput capabilities of our publishing pipeline, i.e. we picked 350 pushes per hour based on historical data and load testing of our publishing pipeline.

If you expect a staged release to do all of the publish at once, and go from completely invisible packages to normally published immediately, this introduces two challenges to the package source:
1. You need to transactionally move N packages from "staged" to "available"
2. This eliminates much of the usefulness for upload throttling since you can upload to a stage as slowly as you want, building a massive queue of packages, and the publish in one moment creating a huge spike in real publishes at that moment.

There are middle grounds where the package is available on the [Package content resource](https://learn.microsoft.com/en-us/nuget/api/package-base-address-resource) but not the [Package metadata resource](https://learn.microsoft.com/en-us/nuget/api/registration-base-url-resource) which could perhaps give NuGet.org's publishing pipeline more breathing room, but that all would be costly to investigate. 

I am assuming here that staged packages go through our asynchronous validation pipeline (malware scanning, repository signing) prior to release. This would be necessary for a timely release since you don't want one package stuck in a slower malware scan to block or break the release. I have no concerns for this from a scalability perspective since our validation pipeline is typically not our bottleneck. It's the publishing pipeline that updates the various V3 API resources.

Author: I think I've covered this by the Asynchronous operations behavior mentioned in the proposal.
